### PR TITLE
mgr/cephadm: Reschedule nfs daemons from offline hosts

### PR DESCRIPTION
**Pull request by adk3798**
In order to improve nfs availability, if there are other
hosts we can place an nfs daemon on or if there is a host
with a lower rank nfs daemon when a higher rank one is on
an offline host, we should reschedule the nfs daemons

Signed-off-by: Adam King <adking@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/cephadm/agent.py [hunk 1]
```diff
@@ -151,6 +150,7 @@ def POST(self) -> Dict[str, Any]:
             # if we got here, we've already verified the keyring of the agent. If
             # host agent is reporting on is marked offline, it shouldn't be any more
             self.mgr.offline_hosts_remove(data['host'])
+            self.mgr.agent_helpers.last_checked_host_offline.pop(data['host'], None)
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/agent.py:153]:
```suggestion
            # This is the last time we checked that the agent on the host is offline
            self.mgr.agent_helpers.last_checked_host_offline.pop(data['host'], None)
```

File: src/pybind/mgr/cephadm/module.py [hunk 2]
```diff
@@ -747,6 +747,17 @@ def _process_ls_output(self, host: str, ls: List[Dict[str, Any]]) -> None:
         self.cache.save_host(host)
         return None
 
+    def check_host_offline(self, host: str, kick_on_failure: bool = False) -> None:
```
Comment (sebastian-philipp) [inline src/pybind/mgr/cephadm/module.py:750]: already done in 

https://github.com/ceph/ceph/blob/b55781d412f05e5ad99751cc4247a22d9ada5547/src/pybind/mgr/cephadm/serve.py#L292-L295

right?

Comment (adk3798) [reply src/pybind/mgr/cephadm/module.py]: yup, just added the reset_con call to that and am now using it

**Comment by adk3798** [comment]
jenkins retest this please

File: src/pybind/mgr/cephadm/ssh.py [hunk 3]
```diff
@@ -75,7 +75,17 @@ async def _remote_connection(self,
...
+                    conn = await asyncio.wait_for(
+                        asyncssh.connect(
+                            addr,
+                            username=self.mgr.ssh_user,
+                            client_keys=[self.mgr.tkey.name],
+                            known_hosts=None,
+                            config=[self.mgr.ssh_config_fname],
+                            preferred_auth=['publickey']
+                        ),
+                        timeout=10
```
[the remark this answers is not in the document]
Comment (adk3798) [reply src/pybind/mgr/cephadm/ssh.py]: Didn't mean to include the commit that had the ssh timeout stuff here

File: src/pybind/mgr/cephadm/agent.py [hunk 4]
```diff
@@ -243,7 +246,19 @@ def handle_metadata(self, data: Dict[str, Any]) -> str:
...
+            for agent in self.mgr.cache.get_daemons_by_type('agent'):
+                assert agent.hostname
+                if agent.hostname not in self.mgr.offline_hosts and self.mgr.agent_helpers._agent_down(agent.hostname):
+                    if agent.hostname not in self.mgr.agent_helpers.last_checked_host_offline:
+                        self.mgr.agent_helpers.last_checked_host_offline[agent.hostname] = datetime.now()
+                    else:
+                        if datetime.now() - self.mgr.agent_helpers.last_checked_host_offline[agent.hostname] <= timedelta(minutes=1):
+                            continue
+                        self.mgr.agent_helpers.last_checked_host_offline[agent.hostname] = datetime.now()
+                    r = CephadmServe(self.mgr)._check_host(agent.hostname)
```
[the remark this answers is not in the document]
Comment (adk3798) [reply src/pybind/mgr/cephadm/agent.py:276]: Really unsure about what to do here in general. What's the best way and time to purely detect a host is offline? Should we implement something for this purpose that does something like a ping?

File: src/pybind/mgr/cephadm/schedule.py [hunk 5]
```diff
@@ -255,6 +256,10 @@ def place(self):
 
         # get candidate hosts based on [hosts, label, host_pattern]
         candidates = self.get_candidates()  # type: List[DaemonPlacement]
+        if self.primary_daemon_type in RESCHEDULE_FROM_OFFLINE_HOSTS_TYPES:
+            # remove unreachable hosts that are not in maintenance so daemons
+            # on these hosts will be rescheduled
+            candidates = self.remove_non_maintenance_unreachable_candidates(candidates)
```
[the remark this answers is not in the document]
Comment (adk3798) [reply src/pybind/mgr/cephadm/schedule.py:262]: added in test_scheduling.py

File: src/pybind/mgr/cephadm/agent.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sebastian-philipp) [reply src/pybind/mgr/cephadm/agent.py:276]: relates to https://tracker.ceph.com/issues/51733

[the remark this answers is not in the document]
Comment (adk3798) [reply src/pybind/mgr/cephadm/agent.py]: As discussed in weekly, now checking the agent's port as a first try (this required making synchronous attempts to message the agent possible which accounts for a lot of the new line changes), then if that fails resetting the ssh connection and trying a check host to fully verify.

**Comment by adk3798** [comment]
http://pulpito.front.sepia.ceph.com/adking-2022-02-11_16:35:44-orch:cephadm-wip-adk-testing-2022-02-10-1637-distro-basic-smithi/

only ingress failures tracked by https://tracker.ceph.com/issues/53807

**Comment by adk3798** [comment]
http://pulpito.front.sepia.ceph.com/adking-2022-02-24_02:51:26-orch:cephadm-wip-adk-testing-2022-02-23-1713-distro-basic-smithi/

Most Failures tracked by:
https://tracker.ceph.com/issues/54389
https://tracker.ceph.com/issues/54304

and a couple upgrade test failures. One where it mysteriously failed redeploying a grafana daemon (the last daemon redeployed during upgrade) and one where it looped on the mgr needing to upgrade itself. Errors don't seem related to the PRs in the run.

**Comment by adk3798** [comment]
http://pulpito.front.sepia.ceph.com/adking-2022-02-28_06:27:32-orch:cephadm-wip-adk-testing-2022-02-27-1813-distro-basic-smithi/

green

File: src/pybind/mgr/cephadm/schedule.py [hunk 6]
```diff
@@ -433,3 +438,14 @@ def get_candidates(self) -> List[DaemonPlacement]:
         final = sorted(ls)
         random.Random(seed).shuffle(final)
         return ls
+
+    def remove_non_maintenance_unreachable_candidates(self, candidates: List[DaemonPlacement]) -> List[DaemonPlacement]:
```
Comment (ajarr) [inline src/pybind/mgr/cephadm/schedule.py:442]: I wasn't able to figure out what hosts are considered to be in maintenance. So wasn't able to also figure out why we include hosts in maintenance mode as candidates.

**Review by ajarr — COMMENTED** [review]
Thanks @adk3798 !  I was trying to follow the changes. Since I'm not very familiar with the cephadm codebase,  I have a couple of questions.

File: src/pybind/mgr/cephadm/schedule.py (hunk 6 above)
Comment (mgfritch) [reply src/pybind/mgr/cephadm/schedule.py:442]: we don't want to re-schedule daemons which are placed on the nodes in maintenance mode ...

File: src/pybind/mgr/cephadm/schedule.py [hunk 7]
```diff
@@ -433,3 +438,14 @@ def get_candidates(self) -> List[DaemonPlacement]:
...
+        in_maintenance: Dict[str, bool] = {}
+        for h in self.hosts:
+            if h.status.lower() == 'maintenance':
+                in_maintenance[h.hostname] = True
+                continue
+            in_maintenance[h.hostname] = False
+        candidates = [c for c in candidates if c.hostname not in [
+            h.hostname for h in self.unreachable_hosts] or in_maintenance[c.hostname]]
```
[the remark this answers is not in the document]
Comment (mgfritch) [reply src/pybind/mgr/cephadm/schedule.py:450]: ... and it's more readable :+1:

**Review by mgfritch — COMMENTED** [review]
can we split off the agent change into a separate PR ?

File: src/pybind/mgr/cephadm/schedule.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (adk3798) [reply src/pybind/mgr/cephadm/schedule.py]: Done

**Comment by adk3798** [comment]
Changelog: Removed all agent related work here so it's just the rescheduling of nfs daemons

File: src/pybind/mgr/cephadm/agent.py [hunk 8]
```diff
@@ -259,7 +263,23 @@ def handle_metadata(self, data: Dict[str, Any]) -> str:
...
+                if agent.hostname not in self.mgr.offline_hosts and self.mgr.agent_helpers._agent_down(agent.hostname):
+                    if agent.hostname not in self.mgr.agent_helpers.last_checked_host_offline:
+                        self.mgr.agent_helpers.last_checked_host_offline[agent.hostname] = datetime.now(
+                        )
+                    else:
+                        if datetime.now() - self.mgr.agent_helpers.last_checked_host_offline[agent.hostname] <= timedelta(minutes=1):
+                            continue
+                        self.mgr.agent_helpers.last_checked_host_offline[agent.hostname] = datetime.now(
+                        )
+                    if not self.mgr.agent_helpers._message_agent(agent.hostname, self.mgr.agent_cache.agent_ports[host], {'counter': self.mgr.agent_cache.agent_counter[host]}, retry=False):
```
[the remark this answers is not in the document]
Comment (adk3798) [reply src/pybind/mgr/cephadm/agent.py]: Removed

File: src/pybind/mgr/cephadm/utils.py [hunk 9]
```diff
@@ -24,6 +24,7 @@ class CephadmNoImage(Enum):
 CEPH_TYPES = ['mgr', 'mon', 'crash', 'osd', 'mds', 'rgw', 'rbd-mirror', 'cephfs-mirror']
 GATEWAY_TYPES = ['iscsi', 'nfs']
 MONITORING_STACK_TYPES = ['node-exporter', 'prometheus', 'alertmanager', 'grafana', 'loki', 'promtail']
+RESCHEDULE_FROM_OFFLINE_HOSTS_TYPES = ['nfs']
```
[the remark this answers is not in the document]
Comment (adk3798) [reply src/pybind/mgr/cephadm/utils.py:27]: Yeah, i think so. Just want to have this PR for only nfs as that's priority

**Comment by adk3798** [comment]
jenkins test api

**Comment by adk3798** [comment]
jenkins test api

**Review by mgfritch — APPROVED** [review]

**Pull request merged**